### PR TITLE
#746 refactored logging of pmd check

### PR DESCRIPTION
**Pull request by krzyk**
#746:
* pmd check logging refactored (it now uses `ResourceValidator`)
* fixed maven and ant plugins and tests

### CODE REVIEW

**Comment by krzyk** [comment]
@davvd please find a reviewer

**Comment by davvd** [comment]
@krzyk Let me find a reviewer for this pull request, thanks for submitting it

**Comment by davvd** [comment]
@mkordas review this one, please

**Comment by davvd** [comment]
> @davvd please find a reviewer

@krzyk OK, got it

**Comment by mkordas** [comment]
@krzyk I'll do a review here during the evening

**Comment by krzyk** [comment]
@mkordas ping

**Comment by mkordas** [comment]
@krzyk sorry, I've missed this one yesterday, on it now :)

File: qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java [hunk 1]
```diff
@@ -45,46 +46,53 @@
...
-            throw new ValidationException(
-                "%d PMD violations (see log above)",
-                size
-            );
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public Collection<Violation> validate(final Collection<File> files) {
+        final SourceValidator validator = new SourceValidator(this.env);
+        final Collection<DataSource> sources = new LinkedList<>();
+        for (final File file : files) {
+            sources.add(new FileDataSource(file));
```
Comment (mkordas) [inline qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java]: @krzyk are you sure that this object creation for each source file won't be a performance problem?

File: qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java [hunk 2]
```diff
@@ -45,46 +46,53 @@
...
         }
-        Logger.info(
-            this,
-            "No PMD violations found in %d files",
-            sources.size()
+        final Collection<RuleViolation> breaches = validator.validate(
+            sources, this.env.basedir().getPath()
         );
+        final Collection<Violation> violations = new LinkedList<>();
```
Comment (mkordas) [inline qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java]: @krzyk can we use more performant `ArrayList` here with initial size set to `breaches.size()`?

File: qulice-maven-plugin/src/it/pmd-violations/verify.groovy [hunk 3]
```diff
@@ -32,8 +32,6 @@
  */
 
 def log = new File(basedir, 'build.log')
-assert log.text.contains('2 PMD violation')
-assert log.text.contains(' PMD violations (see log above)')
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/pmd-violations/verify.groovy]: @krzyk do we have any replacement for these assertions?

File: qulice-maven-plugin/src/it/log-check/verify.groovy [hunk 4]
```diff
@@ -32,8 +32,6 @@
  */
 
 def log = new File(basedir, 'build.log')
-assert log.text.contains('Starting PMD validator')
-assert log.text.contains('Finishing PMD validator')
```
Comment (mkordas) [inline qulice-maven-plugin/src/it/log-check/verify.groovy]: @krzyk why we're not printing these markers anymore for `ResourceValidator`?

File: qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java (hunk 1 above)
Comment (mkordas) [inline qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java]: @krzyk nevermind, I see that you've just moved this method

File: qulice-pmd/src/main/java/com/qulice/pmd/Files.java [hunk 5]
```diff
@@ -1,73 +0,0 @@
...
-        this.environment = environment;
-    }
-
-    /**
-     * Get full list of files to process.
-     * @return Collection of data sources.
-     */
-    public Collection<DataSource> sources() {
-        final Collection<DataSource> sources = new LinkedList<>();
-        for (final File file : this.environment.files("*.java")) {
```
Comment (mkordas) [inline qulice-pmd/src/main/java/com/qulice/pmd/Files.java]: @krzyk where do we limit just to Java files now?

File: qulice-pmd/src/test/java/com/qulice/pmd/FilesTest.java [hunk 6]
```diff
@@ -1,66 +0,0 @@
...
-    /**
-     * Should provide only java files.
-     * @throws IOException In case of problem.
-     */
-    @Test
-    public void providesOnlyJavaFiles() throws IOException {
-        final String source = "class Cls{}";
-        final Environment env = new Environment.Mock()
-            .withFile("src/main/java/Main.java", source)
-            .withFile("src/main/resources/test.properties", "prop=1");
```
Comment (mkordas) [inline qulice-pmd/src/test/java/com/qulice/pmd/FilesTest.java]: @krzyk do we have replacement for a test with non-java file?

**Comment by mkordas** [comment]
@krzyk see my review above

File: qulice-pmd/src/main/java/com/qulice/pmd/Files.java (hunk 5 above)
Comment (krzyk) [inline qulice-pmd/src/main/java/com/qulice/pmd/Files.java]: @mkordas we don't as there is no point in that. PMD and checkstyle have checks for other file types also

File: qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java (hunk 2 above)
Comment (krzyk) [inline qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java]: @mkordas we don't random access, we only iterate over this list

File: qulice-pmd/src/test/java/com/qulice/pmd/FilesTest.java (hunk 6 above)
Comment (krzyk) [inline qulice-pmd/src/test/java/com/qulice/pmd/FilesTest.java]: @mkordas no, we will process non-java files from now on also

File: qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java (hunk 2 above)
Comment (mkordas) [inline qulice-pmd/src/main/java/com/qulice/pmd/PmdValidator.java]: @krzyk `ArrayList` is faster for iteration as well, but here it doesn't matter at all, so let's leave it

File: qulice-pmd/src/main/java/com/qulice/pmd/Files.java (hunk 5 above)
Comment (mkordas) [inline qulice-pmd/src/main/java/com/qulice/pmd/Files.java]: @krzyk indeed... but does it mean that e.g. `pmd-jsp` was not working at all? https://github.com/teamed/qulice/blob/master/qulice-pmd/pom.xml#L81

File: qulice-pmd/src/test/java/com/qulice/pmd/FilesTest.java (hunk 6 above)
Comment (mkordas) [inline qulice-pmd/src/test/java/com/qulice/pmd/FilesTest.java]: @krzyk that's good, PMD has some useful non-java check, but we're not using them

**Comment by mkordas** [comment]
@krzyk I'm merging, as the code looks good, but please take a look at 3 questions that are still open here

**Comment by mkordas** [comment]
@rultor merge

File: qulice-pmd/src/main/java/com/qulice/pmd/Files.java (hunk 5 above)
Comment (krzyk) [inline qulice-pmd/src/main/java/com/qulice/pmd/Files.java]: @mkordas it depends on the rules, if there were any jsp ones (I doubt it) then they wouldn't be used.

**Comment by krzyk** [comment]
@mkordas oops, you went with it too fast, I had some changes to implement based on your comments :)

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by krzyk** [comment]
@rultor deploy pls

File: qulice-pmd/src/main/java/com/qulice/pmd/Files.java (hunk 5 above)
Comment (mkordas) [inline qulice-pmd/src/main/java/com/qulice/pmd/Files.java]: @krzyk we have entire ruleset enabled: https://github.com/teamed/qulice/blob/7525ccf0c14cc525062853844506837f92a66ca1/qulice-pmd/src/main/resources/com/qulice/pmd/ruleset.xml#L119 - looks like now they may start to work, but we don't have any tests to verify the impact - do you think it qualifies for separate issue? We should either remove JSP or have test IMO.

**Comment by mkordas** [comment]
@krzyk 
> oops, you went with it too fast, I had some changes to implement based on your comments :)

You can raise new PR and I'll review it quickly :)

**Comment by krzyk** [comment]
@mkordas yeah I know, but davvd is not so quick and he doesn't work on weekends :), I'll add it to another issue

**Comment by davvd** [comment]
@mkordas since quality is good, 10 mins paid to @krzyk (our architect), transaction `98619891`

I added **32 mins** to your account, many thanks for working with the project (57c535b1968012035236e050)! The completion time here was 70 hours and 20 mins.

review comments (c=17) added as a bonus

added +32 to your rating, now it is equal to [+4428](http://www.netbout.com/b/36080?open=rating)

**Comment by davvd** [comment]
@rultor deploy pls

**Pull request merged**